library/char-stream.lisp

```diff
@@ -0,0 +1,427 @@
+(coalton-library/utils:defstdlib-package #:coalton-library/char-stream
+  (:use #:coalton
+        #:coalton-library/classes)
+  (:local-nicknames (#:gray #:trivial-gray-streams))
+  (:import-from #:coalton-impl/runtime
+                #:function-entry)
+  (:export
+   #:StreamError
+   #:StreamErrorSimple
+   #:StreamErrorClosed
+   #:StreamErrorEndOfFile
+   #:StreamErrorDecoding
+   #:StreamErrorEncoding
+   #:StreamErrorTimeout
+   #:StreamErrorReader
+
+   #:Stream
+   #:open?
+   #:close!
+
+   #:Input
+   #:read-char!
+   #:read-line!
+
+   #:FlushOperation
+   #:FlushBlocking
+   #:FlushAsync
+
+   #:Output
+   #:write-char!
+   #:write-string!
+   #:flush-output!
+
+   #:write-line!
+
+   ;; common lisp condition classes
+   #:coalton-stream-error
+   #:coalton-stream-error-simple
+   #:coalton-stream-error-simple-message
+   #:coalton-stream-error-closed
+   #:coalton-stream-error-end-of-file
+   #:coalton-stream-error-decoding
+   #:coalton-stream-error-encoding
+   #:coalton-stream-error-timeout
+   #:coalton-stream-error-reader))
+
+#+coalton-release
+(cl:declaim #.coalton-impl:*coalton-optimize-library*)
+
+(cl:in-package #:coalton-library/char-stream)
+
+(cl:defmacro %define-into-as-unsafe-coerce (from-type to-type)
+  (alexandria:with-gensyms (instance)
+    `(define-instance (Into ,from-type ,to-type)
+       (define (into ,instance)
+         (lisp ,to-type (,instance)
+           ,instance)))))
+
+(coalton-toplevel
+  ;; users should interact with streams via the classes `Input' and `Output', rather than explicitly
+  ;; referencing these types
+  (repr :native cl:stream)
+  (define-type %AbstractStream)
+  
+  (repr :native (cl:and cl:stream (cl:satisfies cl:input-stream-p)))
```

Comment:
Representing these types these way is really clever. Unfortunately `input-stream-p` starts returning `nil` for file streams once they've been closed, and that was causing me runtime errors when I was experimenting with it. In my branch I had to change these three types to just `(repr :native cl:stream)` to get around this.

It's not as elegant, and there might be a better solution. If we do it that way, we'll have to rely on exposing an API that is very careful to only return these types if they're actually that kind of stream.

Here's a CL sample to demonstrate:
```Common Lisp
(in-package :cl-user)

(defun main ()
  (let ((stream (open "numbers.txt")))
    (print (input-stream-p stream)) ;; Prints T
    (close stream)
    (print (input-stream-p stream)))) ;; Prints nil
```

Reply:
> input-stream-p starts returning nil for file streams once they've been closed

It wHAT? That's absurd. I suspect it's SBCL weirdness and not a portable behavior; I can't find anything about it in the spec. Oh well, there go the cool types. `%InputStream` should definitely contain closed streams, so I guess all of those types will just be `cl:stream`. Sigh.

Reply:
You might interpret this conclusion from a strict reading of the spec:

Glossary:

> input adj. (of a stream) supporting input operations (i.e., being a ``data source''). 

Function INPUT-STREAM-P:

> input-stream-p returns true if stream is an input stream

Function CLOSE:

> close closes stream. Closing a stream means that it may no longer be used in input or output operations.

Reply:
(still bogus tho)